### DISCUSSION TITLE
Using schema as function parameter with safeParse

### DISCUSSION THREAD

**Discussion by drchoppy** [colinhacks/zod#3359] [Q&A]
I'm trying to make a generic 'parse response' function in a database module designed for interacting with DynamoDB:

````
const parseDynamoResponse = <T extends z.ZodTypeAny>(items: Record<string, unknown>[] | undefined, schema: T) => {
    if (!items) return [];
    return items.map((item) => {
        const validatedItem = schema.safeParse(item);
        if (validatedItem.success) {
            return validatedItem.data as z.infer<T>;
        }
        logger().warn("item failed validation", validatedItem);
        return null;
    }).filter(isNotNullOrUndefined);
};
`````

I'm trying to follow the implementation in the docs for using generic type arguments for schemas. However, even with the suggested typecasting  for `return validItem.data as z.infer<T>;`,  TypeScript and ESLint are still not happy with that particular line: `ESLint: Unsafe return of type 'any' from function with return type 'TypeOf<T> | null'.(@typescript-eslint/no-unsafe-return`

Can someone help me out with what I'm doing wrong here?

**Comment by brenc** [comment]
I'm not sure there is any way to fix this. That's what brought me here too. Here's my code:
```typescript
import { parse, type TomlPrimitive } from 'smol-toml';
import { z, ZodError } from 'zod';
import { fromError } from 'zod-validation-error';

export async function loadConfig<TZodType extends z.ZodTypeAny>(
  path: string,
  schema: TZodType,
): Promise<z.infer<TZodType>> {
  let rawConfig: Record<string, TomlPrimitive>;
  try {
    rawConfig = parse(await Bun.file(path).text());
  } catch (err: unknown) {
    if (!(err instanceof Error)) {
      throw err;
    }
    throw new Error(`error parsing config "${path}": ${err.message}`);
  }

  try {
    // There isn't any way to fix this afaik.
    // eslint-disable-next-line @typescript-eslint/no-unsafe-return
    return schema.parse(rawConfig) as z.infer<TZodType>;
  } catch (err: unknown) {
    if (!(err instanceof ZodError)) {
      throw err;
    }
    throw fromError(err);
  }
}
```
TS definitely sees the return type as `z.TypeOf<TZodType>` so I think this is a bug or limitation of eslint. I'm fairly confident that disabling the rule here is the correct solution (but I'm here looking at discussions and bug reports to make sure).

**Reply by drchoppy** [reply]
Yes, disabling ES lint for that line seemed to be the only way of resolving this for us as well. I think you are right, it's more a limitation of ES Lint than Zod. Dependencies don't always play nice, I guess
